packages/eslint-plugin/src/rules/no-redundant-type-constituents.ts

```diff
@@ -87,107 +43,257 @@ function addToMapGroup<Key, Value>(
+      sourceType,
+      targetType,
+      checker,
+    );
+
+    return sourceAssignableToTarget && !targetHasOnlyOptionalProperty;
   }
+  return checker.isTypeAssignableTo(sourceType, targetType);
+}
 
-  if (tsutils.isFalseLiteralType(type)) {
-    return 'false';
+function mergeTypeNames(
+  typeWithNames: TypeWithName[],
+  operator: '&' | '|',
+): string {
+  if (typeWithNames.length === 1) {
+    return typeWithNames[0].typeName;
   }
 
-  return 'literal type';
+  const wrapType = (typeWithName: TypeWithName) => {
+    if (operator === '|' && typeWithName.type.isIntersection()) {
+      return `(${typeWithName.typeName})`;
+    }
+    if (operator === '&' && typeWithName.type.isUnion()) {
+      return `(${typeWithName.typeName})`;
+    }
+    return typeWithName.typeName;
+  };
+
+  return typeWithNames.map(wrapType).join(` ${operator} `);
 }
 
-function describeLiteralTypeNode(typeNode: TSESTree.TypeNode): string {
-  switch (typeNode.type) {
-    case AST_NODE_TYPES.TSAnyKeyword:
-      return 'any';
-    case AST_NODE_TYPES.TSBooleanKeyword:
-      return 'boolean';
-    case AST_NODE_TYPES.TSNeverKeyword:
-      return 'never';
-    case AST_NODE_TYPES.TSNumberKeyword:
-      return 'number';
-    case AST_NODE_TYPES.TSStringKeyword:
-      return 'string';
-    case AST_NODE_TYPES.TSUnknownKeyword:
-      return 'unknown';
-    case AST_NODE_TYPES.TSLiteralType:
-      switch (typeNode.literal.type) {
-        case TSESTree.AST_NODE_TYPES.Literal:
-          switch (typeof typeNode.literal.value) {
-            case 'bigint':
-              return `${typeNode.literal.value < 0 ? '-' : ''}${
-                typeNode.literal.value
-              }n`;
-            case 'string':
-              return JSON.stringify(typeNode.literal.value);
-            default:
-              return `${typeNode.literal.value}`;
-          }
-        case TSESTree.AST_NODE_TYPES.TemplateLiteral:
-          return 'template literal type';
-      }
+function getTypeNodeFromReferenceType(type: ts.Type): ts.Node | undefined {
+  const symbol = type.getSymbol() ?? type.aliasSymbol;
+  const declaration = symbol?.getDeclarations()?.[0];
+
+  if (declaration) {
+    if (ts.isTypeAliasDeclaration(declaration)) {
+      return declaration.type;
+    }
+    return declaration;
   }
+  return;
+}
 
-  return 'literal type';
+function getGroupTypeRelationsByNonRedundantType(
+  typeRedundancyRelations: TypeRedundancyRelation[],
+) {
+  const groups = new Map<ts.Type, TypeRedundancyRelation[]>();
+
+  for (const typeRedundancyRelation of typeRedundancyRelations) {
+    addToMapGroup(
+      groups,
+      typeRedundancyRelation.nonRedundantTypeWithName.type,
+      typeRedundancyRelation,
+    );
+  }
+  return groups;
 }
 
-function isNodeInsideReturnType(node: TSESTree.TSUnionType): boolean {
-  return (
-    node.parent.type === AST_NODE_TYPES.TSTypeAnnotation &&
-    isFunctionOrFunctionType(node.parent.parent)
-  );
+function hasTargetOnlyOptionalProps(
```

Comment:
[Naming] This function name confused me. I _think_ this is what it's actually doing? Did I get it right?

```suggestion
function targetOnlyAddsOptionalProperties(
```